_dev/interop_test_runner.py

```diff
@@ -27,27 +27,47 @@ class Docker(object):
     def __init__(self):
         self.d = docker.from_env()
 
+    def close(self):
+        self.d.close()
+
     def get_ip(self, server):
         tris_localserver_container = self.d.containers.get(server)
         return tris_localserver_container.attrs['NetworkSettings']['IPAddress']
 
     def run_client(self, image_name, cmd):
         ''' Runs client and returns tuple (status_code, logs) '''
-        c = self.d.containers.create(image=image_name, command=cmd)
-        c.start()
+        c = self.d.containers.run(image=image_name, detach=True, command=cmd)
         res = c.wait()
-        ret = c.logs()
+        ret = c.logs().decode('utf8')
         c.remove()
         return (res['StatusCode'], ret)
 
     def run_server(self, image_name, cmd=None, ports=None, entrypoint=None):
         ''' Starts server and returns docker container '''
-        c = self.d.containers.create(image=image_name, detach=True, command=cmd, ports=ports, entrypoint=entrypoint)
-        c.start()
-        # TODO: maybe can be done better?
-        time.sleep(3)
+        c = self.d.containers.run(image=image_name, auto_remove=True, detach=True, command=cmd, ports=ports, entrypoint=entrypoint)
+        self.wait_for_ports(c)
         return c
 
+    def wait_for_ports(self, c):
+        ''' Waits until all services from the image are ready. '''
+        check_ports = set(int(portproto.split('/')[0]) for portproto in c.attrs['NetworkSettings']['Ports'])
+        for i in range(0, 100):
+            # 'ss -Htln' is the modern form, but boringssl image only includes
+            # netstat (via busybox).
+            info = c.exec_run('sh -c "ss -Htln 2>/dev/null || netstat -tln"')
+            if info.exit_code != 0:
+                print('Unable to check for open ports, sleeping for a fixed time')
+                time.sleep(3)
```

Comment:
so you will sleep for 3 seconds and then return anyway?

Reply:
Yes, the old behavior. I'll just add a hard dependency on ss then and remove this general case.